pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java

```diff
@@ -605,30 +615,30 @@ static boolean isValidPropertyValue(String value) {
   @VisibleForTesting
   static String getValidPropertyValue(String value, boolean isMax, DataType dataType) {
     String valueWithinLengthLimit = getValueWithinLengthLimit(value, isMax, dataType);
-    int length = valueWithinLengthLimit.length();
+    String escapedValue = StringEscapeUtils.escapeJava(valueWithinLengthLimit);
+    int length = escapedValue.length();
 
     if (length > 0) {
       // check and replace first character if it's a white space
-      if (Character.isWhitespace(valueWithinLengthLimit.charAt(0))) {
-        String unicodeValue = "\\u" + String.format("%04x", (int) valueWithinLengthLimit.charAt(0));
-        valueWithinLengthLimit = unicodeValue + valueWithinLengthLimit.substring(1);
+      if (Character.isWhitespace(escapedValue.charAt(0))) {
```

Comment:
Will escape handle whitespace? If so, we don't need to explicitly handle whitespace

Reply:
Escape only handles a few of whitespace characters not all. We also need to escape leading and trailing ` `, which is not possible through escape function.

Reply:
Can you give an example of the whitespace that is not skipped? Seems everything < 32 or > 0x7F are skipped

Reply:
Here is my understanding of the problem.

- If the original value has leading and trailing whitespace characters, we must escape them to get the same value from the saved property file. This issue happens because of the String method `trim` usage. Here are the properties-configuration codepoints, reading the entire [line](https://github.com/apache/commons-configuration/blob/RELEASE_1_10_BRANCH/src/main/java/org/apache/commons/configuration/PropertiesConfiguration.java#L748) and value [parse](https://github.com/apache/commons-configuration/blob/RELEASE_1_10_BRANCH/src/main/java/org/apache/commons/configuration/PropertiesConfiguration.java#L927)

- Here is the valid example code for which it fails with the use of escaping as well, 
```java
public static void main(String[] args)
      throws ConfigurationException {
    writeProperty();
    readProperty();
  }

  public static void writeProperty()
      throws ConfigurationException {
    // case 1: fails for leading space, we need to preserve the leading space.
    PropertiesConfiguration configuration = new PropertiesConfiguration(new File("test"));

    String valueWithLeadingSpace = "  valueWithLeadingSpace";
    String escapedValueWithLeadingSpace = StringEscapeUtils.escapeJava(valueWithLeadingSpace);
    System.out.println(escapedValueWithLeadingSpace); // output: '  valueWithLeadingSpace'
    configuration.setProperty("valueWithLeadingSpace", escapedValueWithLeadingSpace);


    // case 2: fails for comma, we need to escape commas
    String valueWithComma = "value,with,comma";
    String escapedValueWithComma =  StringEscapeUtils.escapeJava(valueWithComma);
    System.out.println(escapedValueWithComma); // output: 'value,with,comma'
    configuration.setProperty("valueWithComma", escapedValueWithComma);


    configuration.save();
  }

  public static void readProperty()
      throws ConfigurationException {
    PropertiesConfiguration configuration = new PropertiesConfiguration(new File("test"));

    String valueFromPropertyFile = (String) configuration.getProperty("valueWithLeadingSpace");
    System.out.println(valueFromPropertyFile);  // output: 'valueWithLeadingSpace'
    String unescapedValueWithLeadingSpace = StringEscapeUtils.unescapeJava(valueFromPropertyFile);
    System.out.println(unescapedValueWithLeadingSpace); // output: 'valueWithLeadingSpace'

    String valueWithCommaFromPropertyFile = (String) configuration.getProperty("valueWithComma"); // throws exception here considering it as array
    System.out.println(valueWithCommaFromPropertyFile);
    String unescapedValueWithCommaFromPropertyFile = StringEscapeUtils.unescapeJava(valueWithCommaFromPropertyFile);
    System.out.println(unescapedValueWithCommaFromPropertyFile);
  }
```
For handling the above cases, we are adding the Unicode-based string value of the whitespace character and unescaping it while reading it back.

please let me know, if I am missing something or my interpretation is incorrect.

Reply:
Just found out that properties-configuration is internally also using [escape](https://github.com/apache/commons-configuration/blob/RELEASE_1_10_BRANCH/src/main/java/org/apache/commons/configuration/PropertiesConfiguration.java#L1210) and [unescape](https://github.com/apache/commons-configuration/blob/RELEASE_1_10_BRANCH/src/main/java/org/apache/commons/configuration/PropertiesConfiguration.java#L1376)

Reply:
I did further analysis, are here are my findings
- we only need the unescapeJava in `ColumnMetadataImpl` for converting back the Unicode whitespace character into the original value we are adding [here](https://github.com/apache/pinot/pull/11218/files#diff-f77f250efb4d56b596a7991f65ff00139fb73463111619d64741bec4306efa1aR626) and [here](https://github.com/apache/pinot/pull/11218/files#diff-f77f250efb4d56b596a7991f65ff00139fb73463111619d64741bec4306efa1aR633). For comma it's already handled in properties-configuration enhanced [unescapeJava](https://github.com/apache/commons-configuration/blob/RELEASE_1_10_BRANCH/src/main/java/org/apache/commons/configuration/PropertiesConfiguration.java#L1376) method.

IMO, let's avoid using escapeJava and unescapeJava. We don't need to introduce it just for handling the leading or trailing whitespace. For handling leading or trailing whitespace, here are a couple of options. 

1. Instead of replacing the whitespace character value with its Unicode value, we can add an extra `\` at the start or end. We will remove this additional `/` based on the flag property( say `isWhitespacePreserved`) value(since we are already good with escape flag property)
2. The correct way of handling this is using [custom properties readers and writers](https://commons.apache.org/proper/commons-configuration/userguide/howto_properties.html). However, I think this requires properties-configuration version `2.x`. We have already identified the [work](https://github.com/apache/pinot/issues/11085), so we can wait for the upgrade to be done and drop the handling of whitespace for now.

Reply:
Thanks for taking time to do the research. Great finding that the value is already escaped in the property file.
I tried explicitly escape the `","` by replacing it with `"\\,"`, but it doesn't work if it is preceded by a backslash (I tried replacing `"\\,"` with `"\\\\\\,"` which works in memory, but once the file is saved and read again, it breaks).
To solve both special cases (leading/trailing space and comma), one workaround we can take is to use a preserved character to replace them. Here we can use `"\0"` because it is preserved as the padding character for string values, and string values cannot include it (See `StringUtil.sanitizeStringValue()` where we explicitly remove it).
Implemented the idea in #11223

Reply:
Thanks, closing the pr.